File: Coding-Challenges/graphDistances/graph_distances.py
#!/usr/bin/env python3

class Graph:

    def __init__(self):
        self.vertices = []
        self.adj_list = {}

    def add_vertex(self, v):
        if v not in self.vertices:
            self.vertices.append(v)
            self.adj_list[v] = []

    def add_edge(self, v, w):
        if not self.adj_list.get(v):
            self.add_vertex(v)
        if not self.adj_list.get(w):
            self.add_vertex(w)
        self.adj_list.get(v).append(w)
        self.adj_list.get(w).append(v)


def initialize_color(vertices):
    color = {}
    for i in range(0, len(vertices)):
        color[vertices[i]] = "WHITE"
    return color


def breadth_first_search(graph, start_vertex, callback = None):
    vertices = graph.vertices
    adj_list = graph.adj_list
    color = initialize_color(vertices)
    queue = []

    queue.append(start_vertex)
    while (len(queue) != 0):
        vertex = queue.pop(0)  # gets first element added
        neighbors = adj_list.get(vertex)  # get all self vertex's neighbors
        color[vertex] = "GRAY"  # we have discovered but not completely explored
        for i in range(0, len(neighbors)):
            neighbor = neighbors[i]
            if color[neighbor] == "WHITE":
                color[neighbor] = "GRAY"
                queue.append(neighbor)  # add to queue so we can finish exploring

        # finished exploring self vertex and its adjacent vertices
        color[vertex] = "BLACK"  # mark as completely explored
        if callable(callback):
            callback(vertex)  # execute callback function if there is one


def print_vertex(value):
    print("Visited vertex: " + value)


graph = Graph()

graph.add_vertex("A")
graph.add_edge("A", "B")
graph.add_edge("B", "E")
graph.add_edge("A", "C")
graph.add_edge("A", "D")

vertices = graph.vertices

breadth_first_search(graph, vertices[0], print_vertex)

File: Coding-Challenges/graphDistances/graphDistances.js
#!/usr/bin/env node

class Graph {
  constructor() {
    this.vertices = [];
    this.adjList = new Map();
  }

  addVertex(v) {
    if (!this.vertices.includes(v)) {
      this.vertices.push(v);
      this.adjList.set(v, []);
    }
  }

  addEdge(v, w) {
    if (!this.adjList.get(v)) {
      this.addVertex(v);
    }
    if (!this.adjList.get(w)) {
      this.addVertex(w);
    }
    this.adjList.get(v).push(w);
    this.adjList.get(w).push(v);
  }

  getVertices() {
    return this.vertices;
  }

  getAdjList() {
    return this.adjList;
  }
}


const initializeColor = vertices => {
  const color = {};
  for (let i = 0; i < vertices.length; i++) {
    color[vertices[i]] = "WHITE";
  }
  return color;
}

const breadthFirstSearch = (graph, startVertex, callback) => {
  const vertices = graph.getVertices();
  const adjList = graph.getAdjList();
  const color = initializeColor(vertices);
  const queue = [];

  queue.push(startVertex);
  while (queue.length !== 0) {
    const vertex = queue.shift(); // gets first element added
    const neighbors = adjList.get(vertex); // get all this vertex's neighbors
    color[vertex] = "GRAY"; // we have discovered but not completely explored
    for (let i = 0; i < neighbors.length; i++) {
      const neighbor = neighbors[i];
      if (color[neighbor] === "WHITE") {
        color[neighbor] = "GRAY";
        queue.push(neighbor); // add to queue so we can finish exploring
      }
    }
    // finished exploring this vertex and its adjacent vertices
    color[vertex] = "BLACK"; // mark as completely explored
    if (callback) {
      callback(vertex); // execute callback function if there is one
    }
  }
}

const printVertex = value => {
  console.log("Visited vertex: " + value);
}

const graph = new Graph();

graph.addVertex("A");
graph.addEdge("A", "B");
graph.addEdge("B", "E");
graph.addEdge("A", "C");
graph.addEdge("A", "D");

const vertices = graph.getVertices();

breadthFirstSearch(graph, vertices[0], printVertex);
